File: app/src/main/java/hyunwook/co/kr/raspberrypicamerakotlin/MainActivity.java
package hyunwook.co.kr.raspberrypicamerakotlin;

import android.Manifest;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.SurfaceTexture;
import android.media.Image;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.TextureView;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;

import hyunwook.co.kr.raspberrypicamerakotlin.camera.CameraJava;
import hyunwook.co.kr.raspberrypicamerakotlin.capture.CaptureView;
import hyunwook.co.kr.raspberrypicamerakotlin.capture.CaptureViewer;

/**
 * Created by hyunwook on 2018-07-10.
 */

public class MainActivity extends AppCompatActivity {

    private Handler mBackgroundHandler;
    TextureView textureView;

    static final String TAG = MainActivity.class.getSimpleName();

    Button captureButton;

    Bitmap viewByte;
    private CaptureViewer captureViewer;
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);

        setContentView(R.layout.activity_main);


        if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            return;
        }

        CaptureView captureView = (CaptureView) findViewById(R.id.captureView);
        captureViewer = captureView;



        textureView = (TextureView) findViewById(R.id.texture);
        textureView.setSurfaceTextureListener(this.textureListener);

        captureButton = (Button) findViewById(R.id.capture);
        captureButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                captureButton.setEnabled(false);
                Log.d(TAG, "click..");
                CameraJava.Instance.takePicture(shutterCallback, pictureCallback);

            }
        });
    }

    final CameraJava.ShutterCallback shutterCallback = new CameraJava.ShutterCallback() {
        @Override
        public void onShutter() {
            Log.d(TAG, "Shutter Callback for Camera2");
        }
    };

    final CameraJava.PictureCallback pictureCallback = new CameraJava.PictureCallback() {
        @Override
        public void onPictureTaken(Image image) {
            Log.d(TAG, "Taken picture is here.");
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    captureButton.setEnabled(true);
                }
            });

            ByteBuffer buffer = image.getPlanes()[0].getBuffer();
            byte[] bytes = new byte[buffer.capacity()];
            buffer.get(bytes);

            Bitmap picture = BitmapFactory.decodeByteArray(bytes, 0, bytes.length, null);
            FileOutputStream out = null;

            Log.d(TAG, "bytes -->" + bytes);
            Log.d(TAG, "picture -->" + picture);
            viewByte = Bitmap.createScaledBitmap(picture, 200, 200, true);

            try {
                out = new FileOutputStream(new File(Environment.getExternalStorageDirectory(), "/test.png"));
                picture.compress(Bitmap.CompressFormat.JPEG, 95, out);
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                try {
                    if (out != null) {
                        out.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

        }
    };

    @NotNull
    private final TextureView.SurfaceTextureListener textureListener = new TextureView.SurfaceTextureListener() {
        @Override
        public void onSurfaceTextureSizeChanged(SurfaceTexture p0, int p1, int p2) {

        }

        @Override
        public void onSurfaceTextureUpdated(SurfaceTexture p0) {

        }

        @Override
        public boolean onSurfaceTextureDestroyed(SurfaceTexture p0) {
            return false;
        }

        @Override
        public void onSurfaceTextureAvailable(SurfaceTexture p0, int p1, int p2) {
            CameraJava.Instance.openCamera(getApplicationContext(), textureView, mBackgroundHandler);
        }
    };
}

File: app/src/main/java/hyunwook/co/kr/raspberrypicamerakotlin/MainActivity.kt
package hyunwook.co.kr.raspberrypicamerakotlin

import android.Manifest
import android.content.pm.PackageManager
import android.graphics.SurfaceTexture
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.TextureView
import hyunwook.co.kr.raspberrypicamerakotlin.camera.CameraJava
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    private var mBackgroundHandler : Handler? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //Check Camera Permission.
        if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            //Reboot Raspberry pi, allow Permission android things.
            Log.d(TAG, "No permission...")
            return
        }

        //기존 블랙아이와 동일하기 SurfaceTexture 이용. (광각 + 카메라 반전)
        texture.surfaceTextureListener = textureListener


 /*       capture.setOnClickListener(View.OnClickListener {
            view -> capture();
            true
        })*/

    }

    /*fun capture() {
        var buffer : ByteBuffer = image
    }*/



    //SurfaceTexture interface.
    val textureListener = object : TextureView.SurfaceTextureListener {
        override fun onSurfaceTextureSizeChanged(p0: SurfaceTexture?, p1: Int, p2: Int) {
            //텍스쳐 카메라 버퍼가 바뀔 때..
        }

        override fun onSurfaceTextureUpdated(p0: SurfaceTexture?) {
            //텍스쳐 뷰가 업데이트 될 때
        }

        override fun onSurfaceTextureDestroyed(p0: SurfaceTexture?): Boolean {
            //텍스쳐 뷰가 Destroyed 가 될 때
            return false
        }

        override fun onSurfaceTextureAvailable(p0: SurfaceTexture?, p1: Int, p2: Int) {
            //카메라 오픈 텍스쳐뷰가 이용가능한상태가 되면 (초기)
            CameraJava.Instance.openCamera(this@MainActivity, texture, mBackgroundHandler)
        }
    }
    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}
